### PR TITLE
Register bouncers on container init

### PR DESCRIPTION
**Pull request by TheSpad**
This PR allows users to add bouncers on container init, rather than having to exec into the container and run `cscli`, for automated deployment scenarios. `jq` is added to the Dockerfile to support parsing cscli output.

Supports both environment variables (in the format `BOUNCER_KEY_<NAME>=<API-KEY>`) and docker secrets (in the format `BOUNCER_KEY_<name>` with the contents `<API-KEY>`). Adding multiple bouncers and mixing environment and secrets are supported (though environment will take precedence in the event of conflicting names).

The init script checks that both a name and a key value have been provided, then it checks to see if there is already an existing bouncer with that name registered (in which case it skips it), and then registers the bouncer with the NAME and KEY provided.

This allows you to do something like:
```yml
services:
  crowdsec:
    image: docker.io/crowdsecurity/crowdsec:latest
    container_name: crowdsec
    environment:
      - BOUNCER_KEY_traefik=mysecretkey12345

  bouncer-traefik:
    image: docker.io/fbonalair/traefik-crowdsec-bouncer:latest
    container_name: crowdsec-bouncer-traefik
    environment:
      - CROWDSEC_BOUNCER_API_KEY=mysecretkey12345
    depends_on:
      - crowdsec
```
On first init of the crowdsec container the logs should output something along the lines of:
```
Api key for 'traefik':

   mysecretkey12345

Please keep this key since you will not be able to retrieve it!
```
And `cscli bouncers list` should show:
```
--------------------------------------------------------------
 NAME  IP ADDRESS  VALID  LAST API PULL         TYPE  VERSION 
--------------------------------------------------------------
 traefik             ✔️   2022-03-09T20:55:12Z                
--------------------------------------------------------------
```

### CODE REVIEW

**Comment by zerkms** [comment]
May be `> /dev/null` the output so that secrets didn't appear in logs?

**Comment by TheSpad** [comment]
> May be `> /dev/null` the output so that secrets didn't appear in logs?

I'm of two minds on this but it is probably the right way to go, I just need to add some error handling so that it notifies you properly in the logs if it can't register the bouncer.

File: Dockerfile [hunk 1]
```diff
@@ -13,7 +13,7 @@ RUN SYSTEM="docker" make release
 RUN cd crowdsec-v* && ./wizard.sh --docker-mode && cd -
 RUN cscli hub update && cscli collections install crowdsecurity/linux && cscli parsers install crowdsecurity/whitelists
 FROM alpine:latest
-RUN apk add --no-cache --repository=http://dl-cdn.alpinelinux.org/alpine/edge/community tzdata yq bash && \
+RUN apk add --no-cache --repository=http://dl-cdn.alpinelinux.org/alpine/edge/community tzdata yq jq bash && \
```
Comment (woopstar) [inline Dockerfile:16]: Also required in the Dockerfile.debian maybe?

Comment (TheSpad) [reply Dockerfile:16]: Good catch, the debian dockerfile didn't exist when I did my last PR so I missed it.

**Comment by woopstar** [comment]
README.md file also needs to be updated with the new functionality.

**Comment by he2ss** [comment]
Hi @TheSpad,

Isn't better to have `BOUNCER_NAME` and `BOUNCER_KEY`. So you will avoid additional code just to parse the bouncer name.

**Comment by TheSpad** [comment]
> Hi @TheSpad,
> 
> Isn't better to have `BOUNCER_NAME` and `BOUNCER_KEY`. So you will avoid additional code just to parse the bouncer name.

Then you can't easily add more than one bouncer, unless you do numeric keys or arrays, but in either case it's probably just as much code to parse it out.

File: docker/docker_start.sh [hunk 2]
```diff
@@ -117,6 +117,38 @@ if [ "$DISABLE_POSTOVERFLOWS" != "" ]; then
     cscli -c "$CS_CONFIG_FILE" postoverflows remove $DISABLE_POSTOVERFLOWS
 fi
 
+## Register bouncers via env
+for BOUNCER in $(compgen -A variable | grep -i BOUNCER_KEY); do
+    KEY=$(printf '%s' "${!BOUNCER}")
+    NAME=$(printf '%s' "$BOUNCER" | cut -d_  -f2-)
+    if [[ -n $KEY ]] && [[ -n $NAME ]]; then
+        if ! cscli -c "$CS_CONFIG_FILE" bouncers list -o json | jq -r .[].name | grep -q "${NAME}"; then
```
Comment (he2ss) [inline docker/docker_start.sh:125]: You can refactor this code (125-130) with a function and call it from those 2 contexts.

**Comment by he2ss** [comment]
@TheSpad yes, you're right. I think we can accept this.
I just added one comment and I think it's fine :+1:

**Comment by TheSpad** [comment]
> @TheSpad yes, you're right. I think we can accept this. I just added one comment + if you can add documentation about this env var in the readme (`docker/README.md`), so the readme in the dockerhub also will be updated please.

Done.

**Comment by TheSpad** [comment]
Can I give this a nudge - is there anything else you need me to do to be able to merge this PR?

**Review by he2ss — APPROVED** [review]
Thanks @TheSpad for the contribution :+1:

**Comment by karkazzul** [comment]
Hello,
I have tested your code like this:
```
services:
  crowdsec:
    image: docker.io/crowdsecurity/crowdsec:latest
    container_name: crowdsec
    environment:
      - BOUNCER_KEY_traefik=mysecretkey12345

  bouncer-traefik:
    image: docker.io/fbonalair/traefik-crowdsec-bouncer:latest
    container_name: crowdsec-bouncer-traefik
    environment:
      - CROWDSEC_BOUNCER_API_KEY=mysecretkey12345
      - CROWDSEC_AGENT_HOST=crowdsec:8080
    depends_on:
      - crowdsec
   
```

And then i check the bouncers list in the crowdsec container, I have nothing.

**Comment by TheSpad** [comment]
> Hello, I have tested your code like this:
> 
> ```
> services:
>   crowdsec:
>     image: docker.io/crowdsecurity/crowdsec:latest
>     container_name: crowdsec
>     environment:
>       - BOUNCER_KEY_traefik=mysecretkey12345
> 
>   bouncer-traefik:
>     image: docker.io/fbonalair/traefik-crowdsec-bouncer:latest
>     container_name: crowdsec-bouncer-traefik
>     environment:
>       - CROWDSEC_BOUNCER_API_KEY=mysecretkey12345
>       - CROWDSEC_AGENT_HOST=crowdsec:8080
>     depends_on:
>       - crowdsec
>    
> ```
> 
> And then i check the bouncers list in the crowdsec container, I have nothing.

AFAIK there hasn't been an updated release of the image since this was merged so it won't work yet.

**Comment by karkazzul** [comment]
thanks for your answer i have using the github repos and it's works !
it's just the docker image that's it's not updated

**Comment by karkazzul** [comment]
Do you have an idea when the image on docker hub was updated with these modifications ?

**Comment by TheSpad** [comment]
It'll presumably be whenever they next do a point release, not sure what their release cycle is like.

**Pull request merged**